vcert/connection_tpp.py

```diff
@@ -109,11 +129,220 @@ def auth(self):
+
+    def retrieve_cert(self, certificate_request):
+        log.debug("Getting certificate status for id %s" % certificate_request.id)
+
+        retrive_request = dict(CertificateDN=certificate_request.id, Format="base64", IncludeChain='true')
+
+        if certificate_request.chain_option == "last":
+            retrive_request['RootFirstOrder'] = 'false'
+            retrive_request['IncludeChain'] = 'true'
+        elif certificate_request.chain_option == "first":
+            retrive_request['RootFirstOrder'] = 'true'
+            retrive_request['IncludeChain'] = 'true'
+        elif certificate_request.chain_option == "ignore":
+            retrive_request['IncludeChain'] = 'false'
+        else:
+            log.error("chain option %s is not valid" % certificate_request.chain_option)
+            raise ClientBadData
+
+        status, data = self._post(URLS.CERTIFICATE_RETRIEVE, data=retrive_request)
+        if status == HTTPStatus.OK:
+            pem64 = data['CertificateData']
+            pem = base64.b64decode(pem64)
+            return parse_pem(pem.decode(), certificate_request.chain_option)
+        elif status == HTTPStatus.ACCEPTED:
+            log.debug(data['Status'])
+            return None
+        else:
+            log.error("Status is not %s. %s" % HTTPStatus.OK, status)
+            raise ServerUnexptedBehavior
+
+    def revoke_cert(self, request):
+        if not (request.id or request.thumbprint):
+            raise ClientBadData
+        d = {
+            "Disable": request.disable
+        }
+        if request.reason:
+            d["Reason"] = request.reason
+        if request.id:
+            d["CertificateDN"] = request.id
+        elif request.thumbprint:
+            d["Thumbprint"] = request.thumbprint
+        else:
+            raise ClientBadData
+        if request.comments:
+            d["Comments"] = request.comments
+        status, data = self._post(URLS.CERTIFICATE_REVOKE, data=d)
+        if status in (HTTPStatus.OK, HTTPStatus.ACCEPTED):
+            return data
+        else:
+            raise ServerUnexptedBehavior
+
+    def renew_cert(self, request, reuse_key=False):
+        if not request.id and not request.thumbprint:
+            log.debug("Request id or thumbprint must be specified for TPP")
+            raise CertificateRenewError
+        if not request.id and request.thumbprint:
+            request.id = self.search_by_thumbprint(request.thumbprint)
+        if reuse_key:
+            log.debug("Trying to renew certificate %s" % request.id)
+            status, data = self._post(URLS.CERTIFICATE_RENEW, data={"CertificateDN": request.id})
+            if not data['Success']:
+                raise CertificateRenewError
+            return
+        cert = self.retrieve_cert(request)
+        cert = x509.load_pem_x509_certificate(cert.cert.encode(), default_backend())
+        for a in cert.subject:
+            if a.oid == x509.NameOID.COMMON_NAME:
```

Comment:
switch case will be better here ?

Reply:
native switch-case statement doesn't exist in Python. We can certainly improve the block code with a dictionary, but that change is out of the scope of this PR.